### PR TITLE
Enable leader election in ws-manager-mk2

### PR DESCRIPTION
**Pull request by aledbf**
## Description

Replaces #18419

## Description

Replaces #18419

This introduces an edge case when the maintenance mode is triggered, and we deploy a new version. The standby replica never gets the configmap update. We change the strategy to have one or more stand-by replicas waiting to be the leader and all the replicas watch the configuration configmap.

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at ae677ea</samp>

This pull request adds leader election for the `ws-manager-mk2` component using the Kubernetes API and a `Lease` object. It removes the `--leader-elect` argument from the component and its deployment files, as it is no longer needed. It also reorders some imports in the `sample-workspace` command.

</details>

#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - alerdbf-ha-mk2</li>
	<li><b>🔗 URL</b> - <a href="https://alerdbf-ha-mk2.preview.gitpod-dev.com/workspaces" target="_blank">alerdbf-ha-mk2.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - alerdbf-ha-mk2-gha.15276</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: .github/actions/integration-tests/action.yml [hunk 1]
```diff
@@ -79,6 +79,7 @@ runs:
         TEST_BUILD_ID: ${{ inputs.test_build_id }}
         TEST_BUILD_URL: ${{ inputs.test_build_url }}
         TEST_BUILD_REF: ${{ inputs.test_build_ref }}
+        GOMAXPROCS: "4"
```
Comment (kylos101) [inline .github/actions/integration-tests/action.yml:82]: If this doesn't help, maybe reduce parallel tests here: https://github.com/gitpod-io/gitpod/blob/1404189a9ffb85b221a10d43ada38d0f13660196/test/run.sh#L117

Like from 10 to 4?

File: components/ws-manager-mk2/controllers/workspace_controller.go [hunk 2]
```diff
@@ -500,8 +500,8 @@ func (r *WorkspaceReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		return []string{owner.Name}
 	}
 	err := mgr.GetFieldIndexer().IndexField(context.Background(), &corev1.Pod{}, wsOwnerKey, idx)
-	if err != nil {
-		return err
+	if err != nil && err.Error() == "informer has already started" {
```
Comment (WVerlaek) [inline components/ws-manager-mk2/controllers/workspace_controller.go:503]: interesting, when does this happen?

**Review by WVerlaek — COMMENTED** [review]
looks good, some questions

Also coming back to a question on the previous PR, how do we ensure workspaces don't timeout when a pod gets elected as leader after it had been running for a while?

We use the controller's startup time as the workspace's last activity in https://github.com/gitpod-io/gitpod/blob/3e8e06194a0a8706acc54fbebcf39b31c470b208/components/ws-manager-mk2/pkg/activity/activity.go#L42. If a pod that was standby for 1 hour gets elected, it won't yet have workspace activity stored (as it didn't receive MarkActive requests), so it will think the workspace's last activity was 1 hour ago and timeout the workspace.

We could change `ManagerStartedAt` to e.g. `ControllerActiveAt`, and set this once a pod becomes elected?

File: test/run.sh [hunk 3]
```diff
@@ -114,14 +115,17 @@ if [ "$TEST_SUITE" == "workspace" ]; then
 
   set +e
   # shellcheck disable=SC2086
-  go test -p 10 -v $TEST_LIST "${args[@]}" -parallel-features=true 2>&1  | go-junit-report -subtest-mode=exclude-parents -set-exit-code -out "${RESULTS_DIR}/TEST-${TEST_NAME}.xml" -iocopy
+  go test -p 4 -v $TEST_LIST "${args[@]}" -parallel-features=true -skip-labels="type=maintenance" 2>&1  | go-junit-report -subtest-mode=exclude-parents -set-exit-code -out "${RESULTS_DIR}/TEST-${TEST_NAME}.xml" -iocopy
```
[the remark this answers is not in the document]
Comment (aledbf) [reply test/run.sh:118]: > ports_test.go:81: cannot start workspace: jsonrpc2: code 490 message: You cannot run more than 4 workspaces at the same time. Please stop a workspace before starting another one.
    apis.go:1176: preview status: ready=true, reason=

**Comment by aledbf** [comment]
> Also coming back to a question on the previous PR, how do we ensure workspaces don't timeout when a pod gets elected as leader after it had been running for a while?

This is similar to when we restart ws-manager-mk2 or we deploy a new version. The worst-case scenario is workspaces will run for more time than it should due to the lost state.

File: components/ws-manager-mk2/controllers/workspace_controller.go (hunk 2 above)
Comment (aledbf) [reply components/ws-manager-mk2/controllers/workspace_controller.go]: This happens when you start/stop controllers during a leader election (this is not required anymore)

**Comment by WVerlaek** [comment]
> This is similar to when we restart ws-manager-mk2 or we deploy a new version. The worst-case scenario is workspaces will run for more time than it should due to the lost state.

I don't think it is though, on a restart the `ManagerStartedAt` field also gets reset, but not on leader election. Unless I'm missing something I do believe that pods will timeout after an old standby pod gets elected

File: components/ws-manager-mk2/main.go [hunk 4]
```diff
@@ -153,6 +154,13 @@ func main() {
 
 	activity := activity.NewWorkspaceActivity()
 
+	go func() {
+		for {
+			<-mgr.Elected()
+			activity.ManagerStartedAt = time.Now()
```
Comment (aledbf) [inline components/ws-manager-mk2/main.go:160]: > I don't think it is though, on a restart the ManagerStartedAt field also gets reset, but not on leader election. Unless I'm missing something I do believe that pods will timeout after an old standby pod gets elected

Here

File: test/run.sh [hunk 5]
```diff
@@ -114,14 +115,17 @@ if [ "$TEST_SUITE" == "workspace" ]; then
...
   RC=${PIPESTATUS[0]}
   set -e
 
   if [ "${RC}" -ne "0" ]; then
     FAILURE_COUNT=$((FAILURE_COUNT+1))
   fi
 
+  # shellcheck disable=SC2086
+  go test -v $TEST_LIST "${args[@]}" -labels="type=maintenance" 2>&1
```
[the remark this answers is not in the document]
Comment (aledbf) [reply test/run.sh:127]: done

File: components/ws-manager-mk2/main.go [hunk 6]
```diff
@@ -156,8 +156,15 @@ func main() {
 
 	go func() {
 		for {
-			<-mgr.Elected()
-			activity.ManagerStartedAt = time.Now()
+			select {
+			case <-mgrCtx.Done():
+				return
+			case <-mgr.Elected():
```
Comment (WVerlaek) [inline components/ws-manager-mk2/main.go:162]: i saw that `Elected()` gets closed once elected, it probably doesn't make sense to try to read again from it in the for loop? Or do we exit when we lose election?

Comment (aledbf) [reply components/ws-manager-mk2/main.go:162]: > Or do we exit when we lose election?

I tried that approach, but that means the stand-by replica constantly restarts (never gets ready).
The issue here is that we baked the grpc server in the controller.

Comment (aledbf) [reply components/ws-manager-mk2/main.go:162]: Testing the last commit: if I change the lease using an invalid `holderIdentity`, after the election timeout, the current leader is restarted

**Pull request merged**